### PR TITLE
rpmuncompress: Support -C for zip and 7zip archives

### PR DESCRIPTION
**Pull request by ffesti**
As both don't support an equivalent to tars --strip-components=1 we move the files out of top directory ourselves with shell commands.
    
Support for Ruby gems is still missing - iff at all possible or desirable.

### CODE REVIEW

File: tools/rpmuncompress.c [hunk 1]
```diff
@@ -115,28 +116,28 @@ static int singleRoot(const char *path)
...
 	}
 
+
 	/* Do all entries in the archive start with the same lead directory? */
 	rootLen = sep - rootName + 1;
 	while (archive_read_next_header(a, &entry) == ARCHIVE_OK) {
 	    const char *p = archive_entry_pathname(entry);
 	    if (strncmp(rootName, p, rootLen)) {
-		ret = 0;
+		rootName = _free(rootName);
```
Comment (pmatilai) [inline tools/rpmuncompress.c:129]: Don't free on multiple places, it'll come back to haunt us one way or the other, sooner or later.
AFAICS you could just leave the ret alone and free rootName just before return if ret is anything else than 1?

File: tools/rpmuncompress.c [hunk 2]
```diff
@@ -151,26 +152,39 @@ static char *doUntar(const char *fn)
...
+	 * just extract it into the specified destination path, otherwise
+	 * strip the first path entry and extract in the destination path
+	 */
+	rasprintf(&mkdir, "mkdir '%s' ; ", dstpath);
+	if (needtar)
 	    rasprintf(&stripcd, " -C '%s' %s", dstpath, sr ? "--strip-components=1" : "");
+	else {
+	    char * moveup;
+	    if (sr)
+		rasprintf(&moveup, " ; tmp=`mktemp -u -p'%s'` ; mv %s/%s $tmp ; (shopt -s dotglob; mv $tmp/* %s) ; rmdir $tmp ", dstpath, dstpath, sr, dstpath);
```
Comment (pmatilai) [inline tools/rpmuncompress.c:170]: We know the name of the directory the archive will create, we shouldn't need any globs here? Make a temporary directory, use that as the target directory to extract to (since all the relevant commands seem to support such an option), then rename to the intended location and remove the tmpdir? Assuming that's workable, we could use the same thing for tar and avoid using separate logic for it and the GNU specific --strip-components option. (Updating the tar to use the newly added field was implied in my comment in the ticket, sorry it wasn't entirely clear - always assuming it's possible, BUT our assumptions didn't exactly hold here)

This is all pretty hard to follow though, if we need a shell script then maybe we should write one, or at least make it move obvious by eg placing it in a multiline string section that *looks* like a shell script to begin with :sweat_smile:

**Comment by pmatilai** [comment]
As a side note, this also makes me think that going forward, maybe we should look at using libarchive for this task. https://github.com/libarchive/libarchive/wiki/Examples#user-content-A_Complete_Extractor is much easier to follow than the existing mess of shell fragments sprintf()'ed and strdup()'ed around into a popen() :sweat_smile: But not here and just now.

**Comment by pmatilai** [comment]
Doh sorry, hit the wrong button :facepalm: Need more coffee clearly...

File: tools/rpmuncompress.c [hunk 3]
```diff
@@ -151,26 +152,39 @@ static char *doUntar(const char *fn)
...
+	    rasprintf(&stripcd, "%s'%s'%s", at->dest, dstpath, sr ? moveup : "");
+	    free(moveup);
+	}
+	free(sr);
+    } else {
+	mkdir = strdup("");
+	stripcd = strdup("");
```
Comment (pmatilai) [inline tools/rpmuncompress.c:177]: When in rpm, use rstrdup() aka xstrdup() instead (although in the next weeks we'll need to stop adding more C based stuff in the first place)

File: tools/rpmuncompress.c (hunk 2 above)
Comment (ffesti) [reply tools/rpmuncompress.c:170]: Nah, we need to move the files out of the top level dir. We could extract into the tmpdir - which would be cleaner and saver, but there is no way around moving the files. We can't rename the top level dir to the destination as we can't assume the destination dir is empty or this is the only extraction we do during build.

File: tools/rpmuncompress.c (hunk 1 above)
Comment (ffesti) [reply tools/rpmuncompress.c:129]: While I generally agree that free-ing at multiple places is bad as it is easy to miss a case and end up with memory leaks, this logic does not really apply here. We return either the string we created once or NULL after freeing. There is no way to create a memory leak here as both options are valid.

File: tools/rpmuncompress.c (hunk 2 above)
Comment (ffesti) [reply tools/rpmuncompress.c]: Re-formatted the script part a bit to make it easier to read.

File: tools/rpmuncompress.c (hunk 1 above)
Comment (pmatilai) [reply tools/rpmuncompress.c:129]: It's not about leaking now, it's when you inevitably need to change things later. 
Having a single place that frees what needs freeing is the only sane way to do it while in C.

Comment (pmatilai) [reply tools/rpmuncompress.c:129]: Multiple ways to achieve that of course, if you don't like the integer ret, have another pointer that defaults to NULL and only gets set otherwise when single root is detected (ie "assume failure" pattern), that way you can drop the assignments preceding the gotos too.

File: tools/rpmuncompress.c (hunk 2 above)
Comment (pmatilai) [reply tools/rpmuncompress.c]: Hmm, I didn't think of multiple unpacks to the same -C directory. I don't know how common a scenario that would be but it's a fair point. But, if that's a goal then lets test for it too, because right now it fails:

> $ mkdir /tmp/aabb
$ ./tools/rpmuncompress -x -C /tmp/aabb ../tests/data/SOURCES/test-1.2.3.zip 
mkdir: cannot create directory ‘/tmp/aabb’: File exists

If we care about collisions then we should indeed unpack to a temporary directory initially, it's possible there's a directory that conflicts whereas its contents don't. Should also avoid one of the mv's?

File: tools/rpmuncompress.c [hunk 4]
```diff
@@ -151,26 +155,47 @@ static char *doUntar(const char *fn)
...
+	else {
+	    char * moveup;
+	    if (sr)
+		/* rename top dir to tmp name to avoid collisions */
+		/* then move files within one level up */
+		rasprintf(
+		    &moveup,
+		    " && "
+		    "tmp=`mktemp -u -p'%s'` && "
+		    "mv %s/%s $tmp && "
```
Comment (pmatilai) [inline tools/rpmuncompress.c:179]: Shouldn't we be quoting all these paths, like the mktemp -p is?

File: tools/rpmuncompress.c (hunk 2 above)
Comment (ffesti) [reply tools/rpmuncompress.c]: OK, added -p to the mkdir. Also reworked the whole logic to create the tmpdir first and extract into that.

File: tools/rpmuncompress.c (hunk 4 above)
Comment (ffesti) [reply tools/rpmuncompress.c]: Done.

**Comment by pmatilai** [comment]
Just squash the commits from "Support -C " onwards and we're done - they're all just about adding tests for the thing added there.
